From cbd75e9db675c08a8c2009981332cac7a3cd8c5e Mon Sep 17 00:00:00 2001
From: Huiquan Zhong <huiquan.zhong@intel.com>
Date: Sat, 25 Jul 2015 00:48:22 +0800
Subject: [PATCH] dma: dw: remove dma on/off in suspend/resume callabck

CHT LPDMA is auto power gated with idle indication, and share the same
power rail with LPSS devices, sometimes in resume_early() callback,
LPSS still is power gated, access CFG register would cause system hang.

So remove DMA on/off operation in suspend and resume. And if some device
want to use DMA, turn on DMA with dmaengine_resume() function.

Change-Id: Ib87d44eda6672990bff624288bb9840a34a6fa20
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-38381
Signed-off-by: Huiquan Zhong <huiquan.zhong@intel.com>
---
 drivers/dma/dw/core.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/dma/dw/core.c b/drivers/dma/dw/core.c
index 9eb897b7ec7a..1ece4c5e045b 100644
--- a/drivers/dma/dw/core.c
+++ b/drivers/dma/dw/core.c
@@ -1005,6 +1005,9 @@ static int dwc_control(struct dma_chan *chan, enum dma_ctrl_cmd cmd,
 
 		spin_unlock_irqrestore(&dwc->lock, flags);
 	} else if (cmd == DMA_RESUME) {
+		if (!(dma_readl(dw, CFG) & DW_CFG_DMA_EN))
+			dma_writel(dw, CFG, DW_CFG_DMA_EN);
+
 		if (!dwc->paused)
 			return 0;
 
@@ -1682,7 +1685,6 @@ int dw_dma_suspend(struct dw_dma_chip *chip)
 {
 	struct dw_dma *dw = chip->dw;
 
-	dw_dma_off(dw);
 	if (dw->clk)
 		clk_disable_unprepare(dw->clk);
 
@@ -1696,7 +1698,6 @@ int dw_dma_resume(struct dw_dma_chip *chip)
 
 	if (dw->clk)
 		clk_prepare_enable(dw->clk);
-	dma_writel(dw, CFG, DW_CFG_DMA_EN);
 
 	return 0;
 }
-- 
1.9.1

